asserts/tool/sign.go

```diff
@@ -0,0 +1,249 @@
+	typ := asserts.Type(req.AssertionType)
+	if typ == nil {
+		return nil, fmt.Errorf("invalid assertion type: %q", req.AssertionType)
+	}
+	if req.Revision < 0 {
+		return nil, fmt.Errorf("assertion revision cannot be negative")
+	}
+	if req.AccountKey == nil && req.KeyID == "" {
+		return nil, fmt.Errorf("both account-key and key id were not specified")
+	}
+
+	var nestedStatement nestedStatement
+	err := parseStatement(req, &nestedStatement)
+	if err != nil {
+		return nil, err
+	}
+	if nestedStatement.Headers == nil {
+		// flat headers, reparse
+		err := parseStatement(req, &nestedStatement.Headers)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	headers, err := stringify(nestedStatement.Headers)
+	if err != nil {
+		return nil, err
+	}
+	body := []byte(nestedStatement.Body)
+
+	keyID := req.KeyID
+	authorityID := req.AuthorityID
+
+	if req.AccountKey != nil {
+		if keyID != "" || authorityID != "" {
+			return nil, fmt.Errorf("cannot mix specifying an account-key together with key id and/or authority-id")
+		}
+
+		// use the account-key as a handle to get the information about
+		// signer and key id
+		a, err := asserts.Decode(req.AccountKey)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse handle account-key: %v", err)
+		}
+		accKey, ok := a.(*asserts.AccountKey)
+		if !ok {
+			return nil, fmt.Errorf("cannot use handle account-key, not actually an account-key, got: %s", a.Type().Name)
+		}
+		keyID = accKey.PublicKeyID()
+		authorityID = accKey.AccountID()
+
+		// extra sanity checks about fingerprint
+		pk, err := keypairMgr.Get(authorityID, keyID)
+		if err != nil {
+			return nil, err
+		}
+		expFpr := accKey.PublicKeyFingerprint()
+		gotFpr := pk.PublicKey().Fingerprint()
+		if gotFpr != expFpr {
+			return nil, fmt.Errorf("cannot use found private key, fingerprint does not match account-key, expected %q, got: %s", expFpr, gotFpr)
+		}
+	}
+
+	if authorityID != "" {
+		headers["authority-id"] = authorityID
+	}
+
+	if headers["authority-id"] == "" {
+		return nil, fmt.Errorf("cannot sign assertion with unspecified signer identifier (aka authority-id)")
+	}
+
+	if req.Revision != 0 {
+		headers["revision"] = strconv.Itoa(req.Revision)
+	}
+
+	adb, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		KeypairManager: keypairMgr,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	a, err := adb.Sign(typ, headers, body, keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	return asserts.Encode(a), nil
+}
+
+// stringify lets the invoker use a limited amount of structured input
+// without having to convert everything obvious to strings upfront on
+// their side, we convert integers, bool (to yes|no), list of strings
+// (to comma separated) and nil (to empty).
+func stringify(m map[string]interface{}) (map[string]string, error) {
+	res := make(map[string]string, len(m))
+	for k, w := range m {
+		var s string
+		switch v := w.(type) {
+		case nil:
+			s = ""
+		case bool:
+			if v {
+				s = "yes"
```

Comment:
"true" and "false" are the default terms in Go, Python, Javascript/JSON, etc, so I wouldn't do yes/no.